File: Data Structures & Algorithms/binary-tree-diameter/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
   public int diameterOfBinaryTree(TreeNode root) {
        int[] res = new int[1];  // stores max diameter found
        dia(root, res);
        return res[0];
    }

    public int dia(TreeNode root , int[] res){
        if (root == null) {
            return 0;
        }

        int lHeight = dia(root.left, res);
        int rHeight = dia(root.right, res);

        // Update the result with current diameter at this node
        res[0] = Math.max(res[0], lHeight + rHeight);

        return 1 + Math.max(lHeight, rHeight);
    }
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool same(TreeNode* r1, TreeNode*r2){
        if(!r1 && !r2){
            return true;
        }
        if(r1&& r2 && r1->val==r2->val){
            return same(r1->left,r2->left)&& same(r1->right,r2->right);
        }
        return false;
    }
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(!root ){
            return false;
        }
        if(!subRoot){
            return true;
        }
        if(same(root,subRoot)){
            return true;
        }
        return isSubtree(root->right,subRoot)|| isSubtree(root->left,subRoot);
    }
};
